### PR TITLE
Fixed #34176 -- Fixed grouping by ambiguous aliases.

### PR DESCRIPTION
**Pull request by charettes**
Regression in b7b28c7c189615543218e81319473888bc46d831.

Refs #31377.

Thanks Shai Berger for the report.

### CODE REVIEW

**Comment by charettes** [comment]
Confirmed the generated SQL is the same for `aggregation_regress aggregation queries` tests on SQLite, PostgreSQL, and MySQL.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@charettes Thanks :+1:

**Comment by felixxm** [comment]
@shaib Does it work for you?

**Comment by shaib** [comment]
I'll check later, hopefully tonight (UTC+2)

**Review by shaib — CHANGES_REQUESTED** [review]
(Can't be removed like I wanted)

File: tests/aggregation_regress/tests.py [hunk 1]
```diff
@@ -1827,6 +1827,27 @@ def test_having_subquery_select(self):
         )
         self.assertEqual(set(books), {self.b1, self.b4})
 
+    def test_aggregate_and_annotate_duplicate_columns(self):
+        books = (
+            Book.objects.values("isbn")
+            .annotate(
+                name=F("publisher__name"),
+                num_authors=Count("authors"),
+            )
```
Comment (shaib) [inline tests/aggregation_regress/tests.py:1836]: The test I had in the ticket had two separate annotations to be grouped on; one with `name` as its alias, and one with an original field called `name`. This query has only one field serving both roles, and that does not generate a conflict.

I think this should be made more like I wrote in the ticket:
```suggestion
            Book.objects.values("isbn")
            .annotate(
                name=F("publisher__name"),
                contact_name=F("contact__name"),
                num_authors=Count("authors"),
            )
```

**Review by shaib — CHANGES_REQUESTED** [review]
Hi,

Thanks for the quick attempt, but this doesn't seem to fix things -- what I think is missing in the test is explained inside (if the test in the PR fails without the main code changes, then that's a related problem, but a different one).

**Comment by charettes** [comment]
@shaib could you elaborate on what you mean by _is explained inside_? The test you provided fails on `main` and passes with this branch and `GROUP BY` references is entirely disabled for aliases that are colliding with selected table names.

Here's the SQL that is generated on this branch

```sql
SELECT "aggregation_regress_book"."isbn",
       "aggregation_regress_publisher"."name" AS "name",
       "aggregation_regress_author"."name" AS "contact_name",
       COUNT("aggregation_regress_book_authors"."author_id") AS "num_authors"
FROM "aggregation_regress_book"
INNER JOIN "aggregation_regress_publisher" ON ("aggregation_regress_book"."publisher_id" = "aggregation_regress_publisher"."id")
INNER JOIN "aggregation_regress_author" ON ("aggregation_regress_book"."contact_id" = "aggregation_regress_author"."id")
LEFT OUTER JOIN "aggregation_regress_book_authors" ON ("aggregation_regress_book"."id" = "aggregation_regress_book_authors"."book_id")
GROUP BY "aggregation_regress_book"."isbn",
         "aggregation_regress_publisher"."name",
         "contact_name";
```

**Comment by shaib** [comment]
Sorry, messed up in the first attempt of the comment. Added a second one.

File: tests/aggregation_regress/tests.py (hunk 1 above)
Comment (charettes) [reply tests/aggregation_regress/tests.py:1836]: Adding two annotations is not required to reproduce the issue; the ambiguity is also present if `publisher.name` is selected as it collides with the `name` alias.

We usually add tests represent the minimal set of criteria to reproduce the issue hence why I dropped the `contact_name=F("contact__name")` annotation from the test.

**Comment by shaib** [comment]
... and note also that the `name` field on the main model "does not play" because of the `.vales()` call.

File: tests/aggregation_regress/tests.py (hunk 1 above)
Comment (shaib) [reply tests/aggregation_regress/tests.py:1836]: I just ran the tests in the original project where this came from with this branch, and they failed in the same way as they did originally. Checking further.

**Comment by charettes** [comment]
> ... and note also that the name field on the main model "does not play" because of the .vales() call.

It's not the `name` field on the main model that collides but the related publisher table.

If you do
```
SELECT "aggregation_regress_book"."isbn",
       "aggregation_regress_publisher"."name" AS "name"
FROM "aggregation_regress_book"
INNER JOIN "aggregation_regress_publisher" ON ("aggregation_regress_book"."publisher_id" = "aggregation_regress_publisher"."id")
```

Then a `GROUP BY "name"` will also be considered ambiguous even if the explicit and implicit alias map to the same column. It's the core of the problem, and joining an extra table to throw another alias into the mix seems to only blur what are the minimal set of criteria to reproduce the issue.

In other words

```sql
SELECT "aggregation_regress_book"."isbn",
       "aggregation_regress_publisher"."name" AS "name",
       COUNT("aggregation_regress_book_authors"."author_id") AS "num_authors"
FROM "aggregation_regress_book"
INNER JOIN "aggregation_regress_publisher" ON ("aggregation_regress_book"."publisher_id" = "aggregation_regress_publisher"."id")
LEFT OUTER JOIN "aggregation_regress_book_authors" ON ("aggregation_regress_book"."id" = "aggregation_regress_book_authors"."book_id")
GROUP BY "aggregation_regress_book"."isbn",
         "name"
ORDER BY "aggregation_regress_book"."isbn" ASC
```

Results in

```
django.db.utils.OperationalError: ambiguous column name: name
```

**Comment by shaib** [comment]
@charettes Yep. You're right about this test, but my original tests still failed, so I had to look...

In the original context, the model which carried the field for the annotation (the analog of `Publisher` here) is a Proxy model. Its `_meta.local_concrete_fields` is an empty tuple. I think a similar problem could also surface with MTI -- and since the list of ambiguous names is over-broad anyway, I guess using `concrete_fields` instead should work.

I'll try to prepare a PR on your branch.

**Comment by charettes** [comment]
Ahh I see, I think we'll want to use `_meta.concrete_model.local_concrete_fields` actually.

Concrete MTI won't be an issue as they require a `JOIN` thus an entry in `self.query.alias_map` so there's no point in using `concrete_fields`. Some parent tables that might not be required to be `JOIN`ed against if only leaf/child fields are necessary.

**Comment by shaib** [comment]
@charettes I started with the test, and it did not fail at all in the way I expected. You can see where I stopped (it's a bit late here now) at https://github.com/charettes/django/compare/ticket-34176...shaib:django:ticket-34176?expand=1

At first I just repeated the test with the proxy-using models, and it passed.
Then I added the debug-printing in `compiler.py`, and noticed that the `name` field gets "protected" because of `Author` - although the table is not joined. I don't know what to make of that.

So I tried to use `num_awards`, the other field on `Publisher` -- and when it is added under the proxy, then indeed, it is the alias that is used in the group-by, but surprisingly, that is not an issue.

If I use `num_awards` in the result for `name` in the query, as it is now, things break with "name must appear in the group-by" -- i.e. the alias in the group-by is understood as referring to the column num_awards; if I use an alias that is not also the name of a column, the query is executed successfully. 

I'm confused. Not sure anymore, neither about the right set of tests, nor about the databases' behavior. It does seem to suggest that the original test is affected by the presence of `name` columns on the tables, even though the columns are not queried.

I hope you (or any other interested party) can make sense of it.

**Comment by charettes** [comment]
Latest commit should address your issue with proxy model by using `.concrete_model` with tested coverage without adding more models at the cost of a few bit shuffling in `setUpTestData`.

This all got me wondered if `self.query.alias_map` entries should always be pointed at concrete models to avoid this foot gun; proxy don't mean anything at the query construction level.

**Comment by shaib** [comment]
Hi, 2 points:

First, the way tests are set up in the new commits changes all the past cases to use the FK to a proxy model, breaking them as you noted (and still adding two models, a proxy and an MTI).  I went a different way in https://github.com/shaib/django/commit/8820bc54a18b12dbab4c6b831f962899978c05ec -- adding the proxy model, and an independent model which still uses the same table as the original `Book` model (with a helper-model to also reuse the table of the `Book.authors` M2M). This allowed me to add the test for the proxy model while keeping the test for the non-proxy case (and of course, not affecting any of the other tests); the point of the hackish table-reuse was to avoid duplicating the test-data setup.

Second, and more importantly: I've played a little with the use of aliases in the group-by, and I think I'm making sense of it. It seems the Postgresql implementation, at least is:
  - If a bare name used in a group by uniquely identifies a column, use that
  - Otherwise, if it non-uniquely identifies a column (names a column that appears in more than one table in the query), error out
  - Otherwise, if it names a selected output alias, use that

This fits the PG documentation, in particular: [In case of ambiguity, a GROUP BY name will be interpreted as an input-column name rather than an output column name.](https://www.postgresql.org/docs/15/sql-select.html#SQL-GROUPBY)

MySQL documentation, as far as I could see, does not specify this.

Sqlite docs say the GROUP BY expressions use the same rules as the ORDER BY expressions, and those [prefer aliases to input-column names](https://www.sqlite.org/lang_select.html#orderby).

IMO this means we should do our best to avoid using aliases in the group-by -- or, if we intend to use aliases, use the unambiguous referencing-output-columns-by-index.

Note that on PG, in particular, the use of output aliases as done currently, opens the door to outright wrong results: An unmanaged model could reference a table which has a column with no field, and a query could inadvertently group on that column. That is, e.g.
```python
class M(models.Model):
    class Meta:
        managed = False
        db_table = 'data'
    a = models.IntegerField()
```
But the table `data` also has a column `b`, e.g. a nullable integer, where some records have values (e.g. because, like for many unmanaged models, the table gets filled from some other system). Now consider a query that attempts to group M records according to `a % 10`:
```python
M.objects.annotate(b=Expression(F('a') % Value(10)))
    .values('b')
    .annotate(whatever)
```
If this ends up issuing a 
```sql
SELECT a mod 10 as b, whatever
FROM data JOIN wherever
GROUP BY b
```
Then PG will group on the column `b`, not the expression `b`; and although this may look a little contrived in the abstract, I think we should not accept this possibility.

**Comment by felixxm** [comment]
@charettes Do you have time to keep working on this?

**Comment by charettes** [comment]
@felixxm yes, I should have time tomorrow. I fear we'll have to add a few extra models though which is unfortunate.

**Comment by charettes** [comment]
@felixxm finally got to it, should be ready for a review.

**Comment by felixxm** [comment]
> @felixxm finally got to it, should be ready for a review.

Great, thanks!

@shaib Does it work for you?

**Comment by shaib** [comment]
@charettes @felixxm  Thanks a lot for this! The tests in the original context pass now.

I still think the concern outlined in my last message, about potential silently-wrong query results, is valid (and is not addressed, AFAICT), but I can also see why it wouldn't be considered a high priority.

**Comment by charettes** [comment]
@shaib for your unmanaged example against PG you've provided no matter how I try I always get an "ambiguous alias" message which is effectively a new error/regression but it's not a silent failure like you're reported here. From my investigation on this issue all database are quite vocal when there's a possible ambiguity and none silently fail like you've reported.

Nonetheless I spent time again producing changes that switch all the grouping to be indexed based like you've suggested and I'm quite happy in the reduction of complexity it yields. I also added tests for unmanaged models.

We have to take a decision here though. Since grouping by index is allowed independently of the presence of aliases we can systematically group by selected columns or only do so when an alias is specified (e.g. `annotate(foo=bar)`) which is the reason why group by aliases was introduced in the first place.

I've pushed a commit that systematically groups by selected column alias and a second one that only does so for group by annotations. Both commits seem to pass the suite (with some generated SQL assertion tweaks).

If you have any other concerns *please voice them now* as working on this feature and this particular regression report has been very taxing because problems are reported in phases. First it's an issue why a field shadowing bypass method (that is normally disallowed), then it's due to proxy models, and lastly you mentioned managed models.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by charettes** [comment]
It appears that the feature flag be kept for Oracle as it also doesn't support grouping by selected column index.

**Comment by felixxm** [comment]
@charettes Thanks again :+1:  I'm going to check & merge it today.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Review by felixxm — DISMISSED** [review]

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -2211,40 +2211,25 @@ def set_group_by(self, allow_aliases=True):
...
+            self.values_select = tuple(values_select)
         group_by = list(self.select)
         for alias, annotation in self.annotation_select.items():
             if not (group_by_cols := annotation.get_group_by_cols()):
                 continue
-            if (
-                allow_aliases
-                and alias not in column_names
-                and not annotation.contains_aggregate
-            ):
```
Comment (felixxm) [inline django/db/models/sql/query.py:2247]: I think we need to revert changes in the `set_group_by()` to fix failures on Oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by shaib** [comment]
Hi @charettes ,

First of all, thanks for your hard work on this, and

> If you have any other concerns _please voice them now_ as working on this feature and this particular regression report has been very taxing because problems are reported in phases. First it's an issue with an obscure field shadowing bypass method (that is normally disallowed), then it's due to proxy models, lastly managed models.

I realize how this feels, but let me assure you I do not "sit" on issues. I've said what I found out as I found it out; if I've brought up extra concerns which surprised you, it was because they surprised me first.

> @shaib for your unmanaged example against PG you've provided no matter how I try I always get an "ambiguous alias" message which is effectively a new error/regression but it's not a silent failure like you're reported here. From my investigation on this issue all database are quite vocal when there's a possible ambiguity and none silently fail like you've reported.

A trivial example:

```plpgsql
zorba=> create table M(a integer not null, b integer);
CREATE TABLE
zorba=> insert into M values(1,1);
INSERT 0 1
zorba=> insert into M values(2,11);
INSERT 0 1
zorba=> select * from M;
 a | b  
---+----
 1 |  1
 2 | 11
(2 rows)

zorba=> select a/3 as b from M group by b;
 b 
---
 0
 0
(2 rows)

zorba=> select a/3 as c from M group by c;
 c 
---
 0
(1 row)
```

I don't have time _right now_ to work this into a proper test in the suite, but I'll happily put the time by Sunday.

I'll also review more thoroughly by then.

**Comment by charettes** [comment]
@felixxm 

> I think we need to revert changes in the set_group_by() to fix failures on Oracle.

Judging from [the Oracle failure](https://djangoci.com/job/pull-requests-oracle/561/) I think we actually uncovered a bug in the implementation of 10866a10fe9f0ad3ffdf6f93823aaf4716e6f27c.

If you look at the `test_group_by_subquery_annotation` below it's clear that the explicit grouping by a subquery annotation should not have different behaviour when its alias collides with an involved table columns otherwise wrong results could be returned.

I think the test should be adjusted to represent that here and it should be skipped on Oracle just like other group by subquery tests. I'll submit a commit demonstrating that.

@shaib 

> A trivial example

[This example effectively reproduces a new case of ambiguous aliases](https://dbfiddle.uk/Tmks8e-I) but solely because there's a primary key on `b` which was not part of your theory above. I understand that you're discovering these issues as you go but this is another demonstration of how providing a detailed set of models from the start that reproduces the issues you are encountering could have greatly reduced the time spent on this ticket.

The good news though is that the current solution of grouping by index does address this case as well.

File: tests/aggregation/tests.py [hunk 3]
```diff
@@ -1500,27 +1500,30 @@ def test_aggregation_subquery_annotation_values(self):
...
             .values("rating")
-            .annotate(total_count=Count("rating"))
+            .annotate(total_count=Count("*"))
+            .order_by("rating")
         )
         self.assertEqual(
             list(publisher_qs),
             [
-                {"rating": 4.0, "total_count": 2},
+                {"rating": None, "total_count": 2},
```
Comment (charettes) [inline tests/aggregation/tests.py:1522]: This is the demonstrating that publisher book row get grouped by rating annotation now like they should while previously the query would get incorrectly grouped by `publisher.id`.

I've confirmed that the adjusted test fails with an "ambiguous alias" when applied to 10866a10fe9f0ad3ffdf6f93823aaf4716e6f27c without the `or alias in column_names` branch adjustment which should demonstrate that it remains a proper regression test.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by charettes** [comment]
buildbot, test on oracle.

File: django/db/models/sql/query.py (hunk 2 above)
Comment (charettes) [reply django/db/models/sql/query.py:2247]: Latest commits should address that.

**Comment by shaib** [comment]
Hi, and again, sorry for maybe making it harder than it could be; all I can say is that I, too, have time limits, and I wanted to make sure the case I was thinking of was interesting enough to consider before investing more time in it.

I'd also add that I never intended to imply that Django should behave in a reasonable way when a table's PK is hidden from it. There's a limit to how far that road we should go, I think. The example was just the quickest way to show that silently-wrong results are achievable.

As far as I can tell, the tests added actually do not address what I had in mind; I've added a test that does, in https://github.com/charettes/django/compare/ticket-34176...shaib:django:ticket-34176?expand=1
Luckily, it passes, so -- assuming it is considered valuable -- there's nothing to do other than incorporating it.

**Comment by felixxm** [comment]
> As far as I can tell, the tests added actually do not address what I had in mind; I've added a test that does, in https://github.com/charettes/django/compare/ticket-34176...shaib:django:ticket-34176?expand=1 Luckily, it passes, so -- assuming it is considered valuable -- there's nothing to do other than incorporating it.

Extra tests are always welcome, thanks :+1:
